SMP Packet re-assembly at BT transport level

de-nordic opened the pull request:
The intention of this PR is to add packet re-assembly to SMP.
The PR consists of two commits:
 - adding generic SMP packet re-assembly;
 - adding modification of BT transport that allows to use re-assembly, if requested via Kconfig option.
 

TODO:
 - [x] unit tests of smp_pkasm
 - [x] testing

de-nordic committed:
mgmt/mcumgr: Add optional packet re-assembly to BT transport

New CONFIG_MCUMGR_SMP_BT_PKASM has been added that enabled
SMP packet re-assembly, at transport level, to the BT.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr: Add packet re-assembly support

The commit adds packet re-asembly to SMP that can be used to
collect packet from fragments at SMP transport level.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

carlescufi commented on subsys/mgmt/mcumgr/smp_bt.c:
```diff
@@ -64,10 +114,21 @@ static ssize_t smp_bt_chr_write(struct bt_conn *conn,
+#endif
 }
 
 static void smp_bt_ccc_changed(const struct bt_gatt_attr *attr, uint16_t value)
```
We seem to be using the ccc_changed callback to detect transport connection/disconnection. @MarekPieta @joerchan @cvinayak can we sanity-check this approach to be sure this is the correct one?

joerchan replied:
Should check if this is called when bonded peers are disconnected. I think it is, but not 100% sure.

carlescufi commented on include/mgmt/mcumgr/smp.h:
```diff
@@ -85,6 +85,14 @@ struct zephyr_smp_transport {
 	zephyr_smp_transport_ud_copy_fn *zst_ud_copy;
 	zephyr_smp_transport_ud_free_fn *zst_ud_free;
+
+#ifdef CONFIG_MCUMGR_SMP_PKASM
```
The thing is that, if we want to be consistent with the Bluetooth spec, we should have:

- SMP Segmentatin
- SMP Reassembly

Because the terms "Fragmentation" and  "Recombination" are used in Bluetooth for low-level breaking up in little chunks and recombining them together. In the case of SMP, this happens at a higher layer, so "Segmentation" and "Reassembly" sounds like a better choice.

carlescufi commented:
@philips77 this should ultimately implement https://github.com/apache/mynewt-mcumgr/issues/6

joerchan commented on subsys/mgmt/mcumgr/smp_bt.c:
```diff
@@ -64,10 +114,21 @@ static ssize_t smp_bt_chr_write(struct bt_conn *conn,
 static void smp_bt_ccc_changed(const struct bt_gatt_attr *attr, uint16_t value)
 {
+#ifdef CONFIG_MCUMGR_SMP_PKASM_BT
+	if (value == BT_GATT_ITER_STOP && in_progress) {
```
value isn't BT_GATT_ITER values.

de-nordic replied:
The code here changed significantly.

joerchan commented on subsys/mgmt/mcumgr/smp_bt.c:
```diff
@@ -49,13 +55,57 @@ static ssize_t smp_bt_chr_write(struct bt_conn *conn,
+		struct smp_bt_user_data *ud = zephyr_smp_pkasm_get_ud(&smp_bt_transport);
+
+		if (ud != NULL) {
+			bt_conn_unref(ud->conn);
```
`bt_conn_unref`, should be followed by ud->conn = NULL;

de-nordic replied:
I can add that, but the re-assembly does not care about the contents of the ud and will simply release net_buf to the pool.

de-nordic replied:
Done

joerchan commented on subsys/mgmt/mcumgr/smp_bt.c:
```diff
@@ -64,10 +114,21 @@ static ssize_t smp_bt_chr_write(struct bt_conn *conn,
+	if (value == BT_GATT_ITER_STOP && in_progress) {
+		struct smp_bt_user_data *ud = zephyr_smp_pkasm_get_ud(&smp_bt_transport);
+
+		bt_conn_unref(ud->conn);
```
`bt_conn_unref`, should be followed by ud->conn = NULL;

de-nordic replied:
Done

joerchan commented on subsys/mgmt/mcumgr/smp_bt.c:
```diff
@@ -49,13 +55,57 @@ static ssize_t smp_bt_chr_write(struct bt_conn *conn,
+	 * Collection can only fail due to being unable to allocate new buffer or while receiving
+	 * more data than expected in payload.
+	 */
+	if (in_progress == false && ret >= 0) {
```
`!in_progress`

de-nordic commented:
@joerchan 
> Should check if this is called when bonded peers are disconnected. I think it is, but not 100% sure.

It is but you may verify my. It is also called when you interrupt transmission by, for example, powering down TX device in the middle of transmission.

kapi-no commented on subsys/mgmt/mcumgr/smp_bt.c:
```diff
@@ -49,13 +55,57 @@ static ssize_t smp_bt_chr_write(struct bt_conn *conn,
+	 * more data than expected in payload.
+	 */
+	if (in_progress == false && ret >= 0) {
+		/* On successful collection of first fragment of a new packet, transpor
```
```suggestion
		/* On successful collection of first fragment of a new packet, transport
```

kapi-no commented on subsys/mgmt/mcumgr/smp_bt.c:
```diff
@@ -49,13 +55,57 @@ static ssize_t smp_bt_chr_write(struct bt_conn *conn,
+		 */
+		struct smp_bt_user_data *ud = zephyr_smp_pkasm_get_ud(&smp_bt_transport);
+
+		if (ud != NULL) {
```
```suggestion
		if (ud) {
```

kapi-no commented on subsys/mgmt/mcumgr/smp_bt.c:
```diff
@@ -64,10 +114,21 @@ static ssize_t smp_bt_chr_write(struct bt_conn *conn,
 static void smp_bt_ccc_changed(const struct bt_gatt_attr *attr, uint16_t value)
 {
+#ifdef CONFIG_MCUMGR_SMP_PKASM_BT
+	if (value == BT_GATT_ITER_STOP && in_progress) {
```
You can find possible values in the Bluetooth header - gatt.h:

```
/** Client Characteristic Configuration Values */

/** @def BT_GATT_CCC_NOTIFY
 *  @brief Client Characteristic Configuration Notification.
 *
 *  If set, changes to Characteristic Value shall be notified.
 */
#define BT_GATT_CCC_NOTIFY			0x0001
/** @def BT_GATT_CCC_INDICATE
 *  @brief Client Characteristic Configuration Indication.
 *
 *  If set, changes to Characteristic Value shall be indicated.
 */
#define BT_GATT_CCC_INDICATE			0x0002

```

When the value is set to a different value than **BT_GATT_CCC_NOTIFY**, then the notifications become disabled.

kapi-no commented on subsys/mgmt/mcumgr/smp_pkasm.h:
```diff
@@ -0,0 +1,102 @@
+	return zst->_pkasm.expected;
+}
+
+/**
+ * Pass packet for further processing
+ *
+ * Checks if packet has enough data to be deemed re-assembled and passes it for further processing.
+ * If successful then the re-assembly state in @p zst will indicate that there is no
+ * re-assembly in progress.
+ * The function can be forced to pass data for processing even if packet is not complete,
+ * in which case it is users responsibility to use user data, passed with packet, to notify
+ * receiving end of such case.
+ *
+ * @param zst	the SMP transport;
+ * @param force	process anyway;
+ *
+ * @return	0 on success;
+ *		-EINVAL if there is not re-assembly in progress;
+ *		-ENODATA if there is not enough data to consider packet re-assembled, it has not
+ *		been passed further.
+ */
+int zephyr_smp_pkasm_complete(struct zephyr_smp_transport *zst, bool force);
```
It seems that this API function is unnecessary and its functionality can be incorporated into zephyr_smp_pkasm_collect. This way, we will simplify API for this module and get rid of the EINVAL opcode which can happen due to API misuse.

de-nordic replied:
This is intentionally so that the `collect` caller can decide whether it want's to add something to the "user data" or maybe not pass the frame for completion, for some reason.
Same reason for "force" so that transport could decide if broken frame should be passed up or not.

kapi-no commented on subsys/mgmt/mcumgr/smp_bt.c:
```diff
@@ -41,6 +42,11 @@ static struct bt_uuid_128 smp_bt_svc_uuid = BT_UUID_INIT_128(
 
+/* It there is re-assembly in progress, this should be true. */
+#ifdef CONFIG_MCUMGR_SMP_PKASM_BT
+static bool in_progress;
```
This is already tracked by the PKASM module. You can define a new PKASM API that will tell this layer whether reassembly is ongoing or not.

kapi-no replied:
bump

kapi-no commented on subsys/mgmt/mcumgr/smp_pkasm.c:
```diff
@@ -0,0 +1,105 @@
+int zephyr_smp_pkasm_collect(struct zephyr_smp_transport *zst, const void *buf, uint16_t len)
+{
+	/* No allocated buffer indicates that we are expecting first fragment */
+	if (zst->_pkasm.current == NULL) {
```
Please unify the style of NULL checks in this PR (either _!zst->_pkasm.current_ or like in the line above).

de-nordic replied:
> Please unify the style of NULL checks in this PR (either _!zst->_pkasm.current_ or like in the line above).

I do not think that I have used `!p` where `p` is pointer in any comparison. The NULL, according to C standard is  implementation specific, so it does not have to be 0, which means that neither `if(p)` or `if(!p)` is correct.
Functions returning pointers should either return valid pointer or, which is not recommended by CERT (ERR02-C: Avoid in-band error indicators), NULL, so the comparison should be against NULL.

kapi-no commented on subsys/mgmt/mcumgr/smp_pkasm.c:
```diff
@@ -0,0 +1,105 @@
+	/* No allocated buffer indicates that we are expecting first fragment */
+	if (zst->_pkasm.current == NULL) {
+		if (len >= sizeof(struct mgmt_hdr)) {
+			uint16_t expected = ntohs(((struct mgmt_hdr *)buf)->nh_len);
```
Can we change this to general converter API: _sys_be16_to_cpu_ and get rid of the _net_ip_ header? This won't necessarily have to process IP packets so the chosen header may be misleading.

de-nordic replied:
Done

kapi-no commented on subsys/mgmt/mcumgr/smp_pkasm.h:
```diff
@@ -0,0 +1,102 @@
+extern "C" {
+#endif
+
+struct zephyr_smp_transport;
```
Please add header dependencies here (e.g header with _zephyr_smp_transport_ definition)

de-nordic replied:
I am not sure about this. The re-assembly code is internal SMP thing that is going to be included only to files that already import the mentioned headers, so this will only duplicate the code. The header has been provided as separate unit to make it easier to figure out which set of functions is re-assembly "API".

kapi-no commented on subsys/mgmt/mcumgr/lib/cmd/img_mgmt/src/img_mgmt.c:
```diff
@@ -493,8 +493,9 @@ img_mgmt_upload(struct mgmt_ctxt *ctxt)
 
-		rc = img_mgmt_impl_write_image_data(req.off, req.img_data, action.write_bytes,
-						    last);
+		//rc = img_mgmt_impl_write_image_data(req.off, req.img_data, action.write_bytes,
+		//				    last);
+		rc = 0;
```
Leftover code from tests

de-nordic replied:
Actually I have left it intentionally here, for the draft, because it allows to give flash some slack during tests and removes flash operations delay; when comparing performance against non-re-assembly code, in image upload scenario, this should be applied to remove flash operations impact on testing.

kapi-no commented on subsys/mgmt/mcumgr/smp.c:
```diff
@@ -6,12 +6,16 @@
 #include "smp/smp.h"
 #include "mgmt/mcumgr/smp.h"
 
+#define printk(...)
```
leftover code from tests?

de-nordic replied:
Debug, actually.

de-nordic commented on subsys/mgmt/mcumgr/CMakeLists.txt:
```diff
@@ -7,6 +7,7 @@ zephyr_library_sources_ifdef(CONFIG_MCUMGR_SMP_BT smp_bt.c)
 zephyr_library_sources_ifdef(CONFIG_MCUMGR_SMP_SHELL smp_shell.c)
 zephyr_library_sources_ifdef(CONFIG_MCUMGR_SMP_UART smp_uart.c)
 zephyr_library_sources_ifdef(CONFIG_MCUMGR_SMP_UDP smp_udp.c)
+zephyr_library_sources_ifdef(CONFIG_MCUMGR_SMP_PKASM smp_pkasm.c)
```
I agree on renaming the file, I do not agree to do move of fragmentation with this PR. Topic of this PR is to enable re-assembly and, IMHO, janitorial work like moving the code with no functional changes, will only add to review additional commit that makes visible changes longer for no reason.
Currently the re-assembly is conditional and fragmentation is not, so there is no point to move fragmentation out of smp.c as as it is basically smp TX.

de-nordic commented on subsys/mgmt/mcumgr/smp.c:
```diff
@@ -6,12 +6,16 @@
 
 #include <zephyr.h>
+#include <net/net_ip.h>
+#include "lib/mgmt/include/mgmt/mgmt.h"
```
Neither is needed here anymore.

de-nordic replied:
Fixed

de-nordic committed:
mgmt/mcumgr/lib: Prepare mcumgr smp for logging

Commit adds configuration for logging from mcumgr SMP protocol
source, and registers logging module.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
tests/mgmt/mcumgr/smp_reassembly: Unit tests for re-assembly

The commit adds unit tests for SMP packet re-assembly.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

cvinayak commented on include/mgmt/mcumgr/smp.h:
```diff
@@ -85,6 +85,14 @@ struct zephyr_smp_transport {
+	struct {
+		struct net_buf *current;	/* net_buf used for reassembly */
+		uint16_t expected;		/* expected bytes to come */
+	} _reassembly;
```
why `_reassembly`  with `_`?  can this `struct`  be anonymous.

cvinayak commented on subsys/mgmt/mcumgr/smp.c:
```diff
@@ -11,6 +11,19 @@
+ * demoted to allow overriding them.
+ */
+#ifdef CONFIG_ZTEST
+#define WEAK __attribute__((weak))
```
```suggestion
#define WEAK __weak
```

de-nordic replied:
Wil fix.

de-nordic replied:
Done

kapi-no commented:
@de-nordic, could you address my previous comment: https://github.com/zephyrproject-rtos/zephyr/pull/41306#discussion_r773655773?

de-nordic commented:
> @de-nordic, could you address my previous comment: [#41306 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/41306#discussion_r773655773)?

The link does not seem to point to any comment.

kapi-no commented:
@de-nordic, it's the hidden one. Please expand my review comments so that you can see all of them. Then, the GH should be able to focus on this comment when you click the link.

de-nordic commented:
> @de-nordic, it's the hidden one. Please expand my review comments so that you can see all of them. Then, the GH should be able to focus on this comment when you click the link.

Stupid GH failed to show it. Got screen shot from Filip and fixed it.

philips77 commented on subsys/mgmt/mcumgr/Kconfig:
```diff
@@ -297,6 +297,14 @@ endif
+	bool "Enable packet reassembly in Bluetooth SMP transport"
+	select MCUMGR_SMP_REASSEMBLY
+	help
+	  When enabled the SMP BT transport will collect all MTU fragments,
```
When enabled, the SMP BT transport will buffer and reassemble received partial fragments before passing it for further processing.

de-nordic replied:
Done

philips77 commented on subsys/mgmt/mcumgr/smp_bt.c:
```diff
@@ -156,6 +159,63 @@ static ssize_t smp_bt_chr_write(struct bt_conn *conn,
+	LOG_DBG("collect = %d", ret);
+
+	/*
+	 * Collection can only fail due to being unable to allocate new buffer or while receiving
```
Collection can fail only due to lack of memory or by receiving more data than expected.

de-nordic replied:
Fixed

philips77 commented on subsys/mgmt/mcumgr/smp_bt.c:
```diff
@@ -156,6 +159,63 @@ static ssize_t smp_bt_chr_write(struct bt_conn *conn,
+	 * more data than expected in payload.
+	 */
+	if (!started && ret >= 0) {
+		/* On successful collection of first fragment of a new packet, transport
```
On successful collection of the first fragment of a new packet the transport information is always attached. The "user data" is always allocated with a buffer, so at this point the ud pointer will not be NULL.

de-nordic replied:
Fixed

philips77 commented on subsys/mgmt/mcumgr/smp_bt.c:
```diff
@@ -156,6 +159,63 @@ static ssize_t smp_bt_chr_write(struct bt_conn *conn,
+		ud->conn = bt_conn_ref(conn);
+
+	} else if (ret == -ENOMEM) {
+		/* Failed to collect buffer */
```
Failed to collect the buffer.

de-nordic replied:
Fixed

philips77 commented on subsys/mgmt/mcumgr/smp_reassembly.c:
```diff
@@ -0,0 +1,116 @@
+
+int zephyr_smp_reassembly_collect(struct zephyr_smp_transport *zst, const void *buf, uint16_t len)
+{
+	/* No allocated buffer indicates that we are expecting first fragment */
```
*the* first fragment.

de-nordic replied:
Fixed

philips77 commented on subsys/mgmt/mcumgr/smp_reassembly.c:
```diff
@@ -0,0 +1,116 @@
+			uint16_t expected = sys_be16_to_cpu(((struct mgmt_hdr *)buf)->nh_len);
+
+			/*
+			 * Size of header is not included in length found in header but we count
```
The size of *the* header is not included in its length field, therefore it has to be added.

de-nordic replied:
Fixed

philips77 commented on subsys/mgmt/mcumgr/smp_reassembly.c:
```diff
@@ -0,0 +1,116 @@
+		zst->__reassembly.expected -= len;
+	} else {
+		/*
+		 * Fragment is longer than expected size of fragments to receive and will not
```
Fragment is longer than the expected size and would not fit into the buffer.

de-nordic replied:
Fixed

philips77 commented on subsys/mgmt/mcumgr/smp_reassembly.h:
```diff
@@ -0,0 +1,122 @@
+/**
+ * Pass packet for further processing
+ *
+ * Checks if packet has enough data to be deemed re-assembled and passes it for further processing.
```
deemed?

de-nordic replied:
Fixed

philips77 commented on tests/subsys/mgmt/smp_reassembly/src/main.c:
```diff
@@ -0,0 +1,217 @@
+static struct net_buf *backup;
+
+/* The function is called by zephyr_smp_reassembly_complete to pass completed packet
+ * for further processing; since there is nothing to process the frame and the test
```
..since there is nothing to process, (comma) the frame and the test...

de-nordic replied:
Fixed

philips77 commented on tests/subsys/mgmt/smp_reassembly/src/main.c:
```diff
@@ -0,0 +1,217 @@
+
+/* The function is called by zephyr_smp_reassembly_complete to pass completed packet
+ * for further processing; since there is nothing to process the frame and the test
+ * is not really interested in doing that this stub will only backup buffer pointer
```
...doing that, this stub will...

de-nordic replied:
Fixed

philips77 reviewed (changes requested):
Comments fixed. I have 0 knowledge about the code, so at least I went through the comments and made some noise.

nvlsianpu commented on subsys/mgmt/mcumgr/smp_bt.c:
```diff
@@ -156,6 +159,63 @@ static ssize_t smp_bt_chr_write(struct bt_conn *conn,
+	} else if (ret == -ENOMEM) {
+		/* Failed to collect the buffer */
+		ret = BT_GATT_ERR(BT_ATT_ERR_INSUFFICIENT_RESOURCES);
+	} else if (ret < 0 && ret != -ENOMEM) {
```
```suggestion
	} else if (ret < 0) {
```
ret == -ENOMEM is just the case above

de-nordic replied:
I am not sure how a compiler will order the conditionals and the `ret < 0` also catches `-ENOMEM`

de-nordic replied:
Done

nvlsianpu commented on subsys/mgmt/mcumgr/smp_reassembly.c:
```diff
@@ -0,0 +1,119 @@
+	return zst->__reassembly.expected;
+}
+
+int zephyr_smp_reassembly_complete(struct zephyr_smp_transport *zst, bool force)
```
`force == true` parameter in nowhere used.

de-nordic replied:
It is in line 88; it is basically equivalent to "complete anyway"

cvinayak commented on subsys/mgmt/mcumgr/smp_bt.c:
```diff
@@ -156,6 +159,63 @@ static ssize_t smp_bt_chr_write(struct bt_conn *conn,
+		ret = BT_GATT_ERR(BT_ATT_ERR_VALUE_NOT_ALLOWED);
+	}
+
+	/* No more bytes are expected for this packet */
+	if (ret == 0) {
+		zephyr_smp_reassembly_complete(&smp_bt_transport, false);
+		ret = len;
+	}
+
+	/* BT expects entire len to be consumed so this either be error or len */
+	return ret >= 0 ? len : ret;
```
```suggestion
	/* No more bytes are expected for this packet */
	if (ret == 0) {
		zephyr_smp_reassembly_complete(&smp_bt_transport, false);
		return len;
	}

	/* BT expects entire len to be consumed so this either be error or len */
	return ret > 0 ? len : ret;
```

de-nordic replied:
Change the exit logic of the function. Now there is only `return len`.

cvinayak commented on subsys/mgmt/mcumgr/smp_bt.c:
```diff
@@ -175,10 +235,20 @@ static ssize_t smp_bt_chr_write(struct bt_conn *conn,
+	if (zephyr_smp_reassembly_expected(&smp_bt_transport) >= 0 && value == 0) {
+		struct smp_bt_user_data *ud = zephyr_smp_reassembly_get_ud(&smp_bt_transport);
+
+		bt_conn_unref(ud->conn);
```
```suggestion
		bt_conn_unref(ud->conn);
		ud->conn = NULL;
```

de-nordic replied:
Done

cvinayak commented on subsys/mgmt/mcumgr/smp_reassembly.h:
```diff
@@ -0,0 +1,122 @@
+ *
+ * @param zst	the SMP transport.
+ *
+ * Note: for efficiency there is not NULL check on @p zst pointer and it is caller's responsibility
+ * to validate pointer before passing it to this function.
```
```suggestion
 * Note: for efficiency there is no NULL check on @p zst pointer and it is caller's responsibility
 * to validate pointer before passing it to this function.
```

de-nordic replied:
Done

cvinayak commented on subsys/mgmt/mcumgr/smp_reassembly.h:
```diff
@@ -0,0 +1,122 @@
+ * @param buf	buffer with data to add;
+ * @param len	length of data to add;
+ *
+ * Note: For efficiency there are not NULL checks on @p zst and @p buf pointers and it is caller's
+ * responsibility to make sure these are not NULL.  Also @p len should be non-0 as there is no
+ * point to pass empty fragment for re-assembly.
```
```suggestion
 * Note: For efficiency there are no NULL checks on @p zst and @p buf pointers and it is caller's
 * responsibility to make sure these are not NULL.  Also @p len should not be zero as there is no
 * point in passing empty fragment for re-assembly.
```

de-nordic replied:
Done

cvinayak commented on subsys/mgmt/mcumgr/smp_reassembly.h:
```diff
@@ -0,0 +1,122 @@
+ *
+ * @return	number of expected bytes left to complete the packet, 0 means buffer is complete
+ *		and no more fragments are expected;
+ *		-ENOSR if packet length read from header is bigger than CONFIG_MCUMGR_BUF_SIZE,
+ *		which means there is not way to fit it in configured buffer;
```
```suggestion
 *		-ENOSR if packet length read from header is bigger than CONFIG_MCUMGR_BUF_SIZE,
 *		which means there is not way to fit it in configured buffer;
```
```suggestion
 *		-ENOSR if packet length read from header is bigger than CONFIG_MCUMGR_BUF_SIZE,
 *		which means there is no way to fit it in the configured buffer;
```

cvinayak commented on subsys/mgmt/mcumgr/smp_reassembly.h:
```diff
@@ -0,0 +1,122 @@
+ *
+ * @param zst	the SMP transport;
+ *
+ * Note: for efficiency there is not NULL check on @p zst pointer and it is caller's responsibility
+ * to validate pointer before passing it to this function.
```
```suggestion
 * Note: for efficiency there is no NULL check on @p zst pointer and it is caller's responsibility
 * to validate pointer before passing it to this function.
```

de-nordic replied:
Done

cvinayak commented on subsys/mgmt/mcumgr/smp_reassembly.h:
```diff
@@ -0,0 +1,122 @@
+ * @param zst	the SMP transport;
+ * @param force	process anyway;
+ *
+ * Note: for efficiency there is not NULL check on @p zst pointer and it is caller's responsibility
+ * to validate pointer before passing it to this function.
```
```suggestion
 * Note: for efficiency there is no NULL check on @p zst pointer and it is caller's responsibility
 * to validate pointer before passing it to this function.
```

de-nordic replied:
Done

cvinayak commented on subsys/mgmt/mcumgr/smp_reassembly.h:
```diff
@@ -0,0 +1,122 @@
+ *
+ * @return	0 on success and not forced;
+ *		expected number of bytes if forced to complete buffer with not enough data;
+ *		-EINVAL if there is not re-assembly in progress;
```
```suggestion
 *		-EINVAL if there is no re-assembly in progress;
```

de-nordic replied:
Done

cvinayak commented on subsys/mgmt/mcumgr/smp_reassembly.h:
```diff
@@ -0,0 +1,122 @@
+ *
+ * @param zst	the SMP transport;
+ *
+ * Note: for efficiency there is not NULL check on @p zst pointer and it is caller's responsibility
+ * to validate pointer before passing it to this function.
```
```suggestion
 * Note: for efficiency there is no NULL check on @p zst pointer and it is caller's responsibility
 * to validate pointer before passing it to this function.
```

de-nordic replied:
Done

cvinayak commented on subsys/mgmt/mcumgr/smp_reassembly.h:
```diff
@@ -0,0 +1,122 @@
+ *
+ * @param zst	the SMP transport;
+ *
+ * Note: for efficiency there is not NULL check on @p zst pointer and it is caller's responsibility
+ * to validate pointer before passing it to this function.
```
```suggestion
 * Note: for efficiency there is no NULL check on @p zst pointer and it is caller's responsibility
 * to validate pointer before passing it to this function.
```

de-nordic replied:
Done

cvinayak commented on tests/subsys/mgmt/smp_reassembly/Kconfig:
```diff
@@ -0,0 +1,14 @@
+config MCUMGR_SMP_REASSEMBLY_UNIT_TESTS
+	def_bool y
+	select MCUMGR_SMP_REASSEMBLY
+	help
+	  Nothing
```
```suggestion
	help
	  Enable SMP Reassembly unit tests.
```

de-nordic replied:
Done

de-nordic commented on subsys/mgmt/mcumgr/smp_bt.c:
```diff
@@ -156,6 +159,63 @@ static ssize_t smp_bt_chr_write(struct bt_conn *conn,
+
+	} else if (ret == -ENOMEM) {
+		/* Failed to collect the buffer */
+		ret = BT_GATT_ERR(BT_ATT_ERR_INSUFFICIENT_RESOURCES);
```
OK, done return here. Also I have changed the return from function so the overall return from function no longer mixes `len` , `ret` and `BT_GATT_ERR`, now it only returns either `len` or `BT_GATT_ERR` in-band error code.

de-nordic commented on subsys/mgmt/mcumgr/smp_bt.c:
```diff
@@ -156,6 +159,63 @@ static ssize_t smp_bt_chr_write(struct bt_conn *conn,
+
+		zephyr_smp_reassembly_drop(&smp_bt_transport);
+		ret = BT_GATT_ERR(BT_ATT_ERR_VALUE_NOT_ALLOWED);
+	}
```
Since I have changed this to direct `return`, when I add the else I got:
`WARNING:UNNECESSARY_ELSE: else is not generally useful after a break or return`

de-nordic commented:
@cvinayak @joerchan @kapi-no @philips77 @nvlsianpu I need to call you back to the review because I have quite modified `smp_bt_chr_write` , after @cvinayak has reviewed the code and pointed out possible improvements; the changes are: 
 - the error processing is done first and function returns immediately from error paths with `BT_GATT_ERR` type value;
 - check for positive `ret` has been as code for `ret >= 0` has been moved after error processing, so it is no longer needed;
 - final return value, for positive outcome, is reduced to `len`.

joerchan commented:
@de-nordic I've added @alwa-nordic as a reviewer, since he has taken over for me.
